app/code/Magento/Customer/Test/Unit/Ui/Component/Listing/Column/GroupActionsTest.php

```diff
@@ -0,0 +1,253 @@
+        $this->urlBuilderMock->expects($this->any())
+            ->method('getUrl')
+            ->willReturnMap(
+                [
+                    ['customer/group/edit', ['id' => 1], 'http://magento.com/customer/group/edit'],
+                    ['customer/group/delete', ['id' => 1], 'http://magento.com/customer/group/delete']
+                ]
+            );
+
+        $dataSource = $this->component->prepareDataSource($dataSource);
+        $this->assertEquals($expectedDataSource, $dataSource);
+    }
+
+    /**
+     * Test data source with a default customer group
+     *
+     * @dataProvider customerGroupsDataProvider
+     */
+    public function testPrepareDataSourceWithDefaultGroup()
+    {
+        $isDefaultGroup = true;
+        $dataSource = [
+            'data' => [
+                'items' => [
+                    [
+                        'customer_group_id' => 1,
+                        'customer_group_code' => 'General',
+                    ],
+                    [
+                        'customer_group_id' => 0,
+                        'customer_group_code' => 'Not Logged In',
+                    ],
+                ]
+            ]
+        ];
+        $expectedDataSource = [
+            'data' => [
+                'items' => [
+                    [
+                        'customer_group_id' => 1,
+                        'customer_group_code' => 'General',
+                        'name' => [
+                            'edit' => [
+                                'href' => 'http://magento.com/customer/group/edit',
+                                'label' => __('Edit'),
+                                '__disableTmpl' => true,
+                            ]
+                        ]
+                    ],
+                    [
+                        'customer_group_id' => 0,
+                        'customer_group_code' => 'Not Logged In',
+                        'name' => [
+                            'edit' => [
+                                'href' => 'http://magento.com/customer/group/edit',
+                                'label' => __('Edit'),
+                                '__disableTmpl' => true,
+                            ]
+                        ]
+                    ]
+                ]
+            ]
+        ];
+
+        $this->groupManagementMock->expects($this->any())
+            ->method('isReadonly')
+            ->willReturn($isDefaultGroup);
+        $this->escaperMock->expects($this->any())
+            ->method('escapeHtml')
+            ->willReturnMap(
+                [
+                    ['General', null, 'General'],
+                    ['Not Logged In', null, 'Not Logged In']
+                ]
+            );
+        $this->urlBuilderMock->expects($this->any())
+            ->method('getUrl')
+            ->willReturnMap(
+                [
+                    ['customer/group/edit', ['id' => 1], 'http://magento.com/customer/group/edit'],
```

Comment:
Replace magic numbers with `const` named for example - `STUB_LOGGEDIN_CUSTOMER_GROUP_ID`
Instead of literal with sample domain - try to also replace it with `const` - eg. `STUB_GROUP_EDIT_URL`